docs/contributors/coding-guidelines.md

```diff
@@ -201,6 +201,246 @@ alert( 'My name is ' + name + '.' );
+ * @property {string} attributeKey A block attribute key.
+ * @property {number} offset       An attribute value offset, based on the rich
+ *                                 text value.
+ */
+```
+
+Custom types can also be beneficial when describing a set of predefined options. While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids a situation where aligning tags while still respecting a maximum line length of 80 characters becomes challenging due to the space occupied by a literal values union.
+
+```js
+/**
+ * Named breakpoint sizes.
+ * 
+ * @typedef {"huge"|"wide"|"large"|"medium"|"small"|"mobile"} WPBreakpoint
+ */
+```
+
+### Importing and Exporting Types
+
+Use the [TypeScript `import` function](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#import-types) to import type declarations from other files or third-party dependencies.
+
+Since an imported type declaration can occupy an excess of the available line length and become verbose when referenced multiple times, you are encouraged to create an alias of the external type using a `@typedef` declaration at the top of the file, immediately following [the `import` groupings](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports).
+
+```js
+/** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */
+```
+
+Note that all custom types defined in another file can be imported.
+
+When considering which types should be made available from a WordPress package, the `@typedef` statements in the package's entrypoint script should be treated as effectively the same as its public API. It is important to be aware of this, both to avoid unintentionally exposing internal types on the public interface, and as the means by which to expose the public types of a project.
+
+```js
+// packages/data/src/index.js
+
+/** @typedef {import('./registry').WPDataRegistry} WPDataRegistry */
+```
+
+In this snippet, the `@typedef` will support the usage of the previous example's `import('@wordpress/data')`.
+
+#### External Dependencies
+
+Many third-party dependencies will distribute their own TypeScript typings. For these, the `import` semantics should "just work".
+
+![Working Example: `import` type](https://user-images.githubusercontent.com/1779930/70167742-62198800-1695-11ea-9c21-82a91d4a60e2.png)
+
+If you use a [TypeScript integration](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Editor-Support) for your editor, you can typically see that this works if the type resolves to anything other than the fallback `any` type.
+
+For packages which do not distribute their own TypeScript types, you are welcomed to install and use the [DefinitelyTyped](http://definitelytyped.org/) community-maintained types definitions, if one exists.
+
+### Record Types
```

Comment:
The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.

At the same time, I like the idea of encouraging custom types to be defined in the file to make it easier to follow.

Reply:
> The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.

Yeah, I'll see what I can do to improve. I agree especially on the syntax for functions. I was also hesitant about even mentioning custom generic types. I think it could be okay as it is, in that it doesn't go into too much depth, but includes a relevant link for those who might be interested in using them.

Reply:
> > The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.
> 
> Yeah, I'll see what I can do to improve. I agree especially on the syntax for functions. I was also hesitant about even mentioning custom generic types. I think it could be okay as it is, in that it doesn't go into too much depth, but includes a relevant link for those who might be interested in using them.

Added some references in def916547, and changed the text around generic types to try to clarify that it's intended for advanced use-cases.